src/graph_generator.cc

```diff
@@ -47,35 +47,33 @@ namespace detail {
 		case task_type::host_compute:
 		case task_type::device_compute:
 		case task_type::master_node: generate_independent_execution_commands(tsk); break;
+		case task_type::fence: generate_fence_commands(tsk); break;
 		}
 
 		for(auto& t : transformers) {
 			t->transform_task(tsk, m_cdag);
 		}
 
-		// Only execution tasks can have data requirements or reductions
-		if(tsk.get_execution_target() != execution_target::none) {
```

Comment:
Do you think the overhead of running `process_task_data_requirements` on every horizon is negligible?

Reply:
No way to know without profiling, but I've added a shortcut branch around it for horizons and epochs.